presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveCrossJoinWithConstantInput.java

```diff
@@ -0,0 +1,145 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.metadata.FunctionAndTypeManager;
+import com.facebook.presto.spi.plan.FilterNode;
+import com.facebook.presto.spi.plan.JoinType;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.ProjectNode;
+import com.facebook.presto.spi.plan.ValuesNode;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.relational.RowExpressionDeterminismEvaluator;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.SystemSessionProperties.isRemoveCrossJoinWithConstantSingleRowInputEnabled;
+import static com.facebook.presto.sql.planner.PlannerUtils.addProjections;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+/**
+ * When one side of a cross join is one single row of constant, we can remove the cross join and replace it with a project.
+ * <pre>
+ * - Cross Join
+ *      - table scan
+ *          left_field
+ *      - values // only one row
+ *          right_field := 1
+ * </pre>
+ * into
+ * <pre>
+ * - project
+ *      left_field := left_field
+ *      right_field := 1
+ *      - table scan
+ *          left_field
+ * </pre>
+ */
+public class RemoveCrossJoinWithConstantInput
+        implements Rule<JoinNode>
+{
+    private final RowExpressionDeterminismEvaluator rowExpressionDeterminismEvaluator;
+
+    public RemoveCrossJoinWithConstantInput(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.rowExpressionDeterminismEvaluator = new RowExpressionDeterminismEvaluator(functionAndTypeManager);
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return join().matching(x -> x.getType().equals(JoinType.INNER) && x.getCriteria().isEmpty());
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRemoveCrossJoinWithConstantSingleRowInputEnabled(session);
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        PlanNode singleValueInput;
+        PlanNode joinInput;
+        if (isOutputSingleConstantRow(context.getLookup().resolve(node.getRight()), context)) {
+            singleValueInput = context.getLookup().resolve(node.getRight());
```

Comment:
is `context.getLookup().resolve(..)` method call cheap.? Asking since we first do the call in if condition and then again in the block.
Maybe a small nit is to move the `context.getLookup().resolve(..)` calls up since you need to call it for both left and right regardless of which if/else block you are in.